sound/soc/sof/intel/hda-stream.c

```diff
@@ -152,20 +152,39 @@ int hda_dsp_stream_spib_config(struct snd_sof_dev *sdev,
 
 /* get next unused stream */
 struct hdac_ext_stream *
-hda_dsp_stream_get(struct snd_sof_dev *sdev, int direction)
+hda_dsp_stream_get(struct snd_sof_dev *sdev, int direction, int flag)
 {
 	struct hdac_bus *bus = sof_to_bus(sdev);
 	struct hdac_ext_stream *stream = NULL;
 	struct hdac_stream *s;
 
 	spin_lock_irq(&bus->reg_lock);
 
-	/* get an unused stream */
-	list_for_each_entry(s, &bus->stream_list, list) {
-		if (s->direction == direction && !s->opened) {
-			s->opened = true;
-			stream = stream_to_hdac_ext_stream(s);
-			break;
+	/*
+	 * Reserved host dma only stream needs to be allocated from tail of
+	 * list which is organized with stream tag value increased from one to
+	 * max, so that normal capture stream could get the same stream tag
+	 * from host and link dma allocator, otherwise, when applying coupled
+	 * and decoupled mode with the mismatched host and link dma stream
+	 * tag, a capture stream could break another capture stream for it would
+	 * couple link dma channel with host dma channel used by different
+	 * catpure streams.
+	 */
+	if (flag & HDA_STREAM_ALLOC_RESERVED) {
```

Comment:
@RanderWang an alternate suggestion, trace dma channel will always be allocated even before tplg is loaded. So there is no possibility of a stream being open at this point. Why bother iterating then? Cant we just allocate list_last_entry?

Reply:
Yes, I also considered this solution. But there will be another reserved stream for buffer probing which is used for checking buffer content between each processing component in FW, we have to allocate list_second_last_entry.

Reply:
@RanderWang thanks. can you please educate me as to where this buffer probing is done?

Reply:
@keyonjie can you help to answer this question ? Thanks!

Reply:
@ranj063 the FW buffer probing feature is in our plan, we discussed that when in Gdansk @plbossart also there, it is used to probe buffers between components, dump contents to host side for debugging purpose, it will work similar to dma trace.

So here, we need to make sure our solution is extendable, there might be > 1 dailess capture streams in the future.